raft/src/main/java/org/apache/kafka/raft/internals/SerdeRecordsIterator.java

```diff
@@ -0,0 +1,231 @@
+    private boolean isClosed = false;
+
+    public SerdeRecordsIterator(
+        Records records,
+        RecordSerde<T> serde,
+        BufferSupplier bufferSupplier,
+        int maxBatchSize
+    ) {
+        this.records = records;
+        this.serde = serde;
+        this.bufferSupplier = bufferSupplier;
+        this.maxBatchSize = maxBatchSize;
+    }
+
+    @Override
+    public boolean hasNext() {
+        checkIfClosed();
+
+        if (!nextBatch.isPresent()) {
+            nextBatch = nextBatch();
+        }
+
+        return nextBatch.isPresent();
+    }
+
+    @Override
+    public Batch<T> next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException("Batch iterator doesn't have any more elements");
+        }
+
+        Batch<T> batch = nextBatch.get();
+        nextBatch = Optional.empty();
+
+        return batch;
+    }
+
+    @Override
+    public void close() {
+        isClosed = true;
+        allocatedBuffer.ifPresent(bufferSupplier::release);
+        allocatedBuffer = Optional.empty();
+    }
+
+    private void checkIfClosed() {
+        if (isClosed) {
+            throw new IllegalStateException("Serde record batch itererator was closed");
+        }
+    }
+
+    private Optional<Iterator<MutableRecordBatch>> nextBatches() {
+        int recordSize = records.sizeInBytes();
+        if (bytesRead < recordSize) {
+            final MemoryRecords memoryRecords;
+            if (records instanceof MemoryRecords) {
+                bytesRead = recordSize;
+                memoryRecords = (MemoryRecords) records;
+            } else if (records instanceof FileRecords) {
+                final ByteBuffer buffer;
+                if (allocatedBuffer.isPresent()) {
+                    buffer = allocatedBuffer.get();
+                    buffer.compact();
+
+                    if (!buffer.hasRemaining()) {
+                        // The buffer is not big enough to read an entire batch
+                        throw new IllegalStateException(
+                            String.format(
+                                "Unable to read batch from file records buffer %s with maximum batch %s and record size %s",
+                                buffer,
+                                maxBatchSize,
+                                records.sizeInBytes()
+                            )
+                        );
+                    }
+                } else {
+                    buffer = bufferSupplier.get(Math.min(maxBatchSize, records.sizeInBytes()));
+                    allocatedBuffer = Optional.of(buffer);
+                }
+
+                int start = buffer.position();
+                try { 
+                    ((FileRecords) records).readInto(buffer, bytesRead);
+                } catch (IOException e) {
+                    throw new RuntimeException("Failed to read records into memory", e);
+                }
+
+                bytesRead += buffer.limit() - start;
```

Comment:
Maybe I am missing it, but how does this ensure proper alignment? The batch at the end of the buffer could be incomplete, no?

Reply:
@hachikuji are you talking about the buffer passed to MemoryRecords and bytes that could be left over after iterating through the batches?

Reply:
There are a couple of important points here:

1. To `MemoryRecords` we pass a slice.
2. We update the `allocatedBuffer::position` when a batch is successfully read here https://github.com/apache/kafka/pull/10085/files/2dd49397a1848e0b9b9c11aa72d5455c5fccfc3e#diff-9ecaf1dc29d739f7a29639b070051e3750f49ea0ac73dc5c27680dfb8f7b9ef6R156-R157
3. The buffer is flipped by https://github.com/apache/kafka/pull/10085/files/2dd49397a1848e0b9b9c11aa72d5455c5fccfc3e#diff-9ecaf1dc29d739f7a29639b070051e3750f49ea0ac73dc5c27680dfb8f7b9ef6R130

Reply:
Ok, I think I get it. I missed the call to `compact` above.

Reply:
Yes. The fact that we `compact` in subsequent calls is important since it copies unused byte to the start of the byte buffer.